Celigo.ServiceManager.NetSuite.REST/src/RESTletClient.cs

```diff
@@ -0,0 +1,172 @@
+            return this.SendRequest(HttpMethod.Get, requestUrl, authHeader, _emptyContent);
+        }
+
+        public Task<HttpResponseMessage> Post<T>(in string account, 
+                                                 in string token, 
+                                                 in string tokenSecret, 
+                                                 in T message, 
+                                                 params (string key, string value)[] queryParams)
+        {
+            var urlBuilder = this.CreateUrlBuilder(account);
+            string authHeader = this.GetTbaAuthHeader("POST", account, token, tokenSecret, urlBuilder.ToString(), queryParams);
+
+            var requestUrl = this.CreateRequestUrl(urlBuilder, queryParams);
+            return this.SendRequest(HttpMethod.Post, requestUrl, authHeader, CreateJsonMessageContent(message));
+        }
+
+        public Task<HttpResponseMessage> Get(in Passport passport, 
+                                             params (string key, string value)[] queryParams)
+        {
+            ValidateBasicCreds(passport);
+
+            string authHeader = this.GetBasicAuthHeaderValue(passport);
+            Uri requestUrl = this.CreateRequestUrl(this.CreateUrlBuilder(passport.Account), queryParams);
+            
+            return this.SendRequest(HttpMethod.Get, requestUrl, authHeader, _emptyContent);
+        }
+
+        public Task<HttpResponseMessage> Post<T>(in Passport passport,
+                                                 in T message, 
+                                                 params (string key, string value)[] queryParams)
+        {
+            ValidateBasicCreds(passport);
+
+            string authHeader = this.GetBasicAuthHeaderValue(passport);
+            Uri requestUrl = this.CreateRequestUrl(this.CreateUrlBuilder(passport.Account), queryParams);
+            
+            return this.SendRequest(HttpMethod.Post, requestUrl, authHeader, CreateJsonMessageContent(message));
+        }
+
+        private StringBuilder CreateUrlBuilder(string account) => new StringBuilder("https://")
+                                                                    .Append(account.ToLowerInvariant().Replace('_', '-'))
+                                                                    .Append(_restletRelativePath);
+
+        private Uri CreateRequestUrl(StringBuilder urlBuilder, (string key, string value)[] queryParams)
+        {
+            urlBuilder
+                .Append("?script=")
+                .Append(_restlet.Script)
+                .Append("&deploy=")
+                .Append(_restlet.Deploy);
+
+            for (int i = queryParams.Length - 1; i >= 0; i--)
+            {
+                urlBuilder.Append("&");
+                urlBuilder.Append(queryParams[i].key);
+                urlBuilder.Append("=");
+                urlBuilder.Append(queryParams[i].value);
+            }
```

Comment:
No particular reason: just that a reverse for loop is trivially faster than an ascending one. No real impact here, just a practice I follow when the order of the loop doesn't matter.

You are right about the `AppendQueryParam` method - it can be done as an inline function here.

Reply:
Hmm.. on second thought: only lines 119-120 can benefit from the method and I wonder if its worth adding a method for that.

Reply:
How about like this? 
We can add a method as written below:
```
private StringBuilder AppendQueryParam(StringBuilder query, (string key, string value) queryParam, bool isFirstParam = false)
{
    if (isFirstParam)
    {
         query.Append("?");
         query.Append($"{queryParam.key}={queryParam.value}");
         return query;
    }
    query.Append($"&{queryParam.key}={queryParam.value}");
    return query;
}
```
And then we can modify the invocation inside `CreateRequestUrl` method as below:
```
private Uri CreateRequestUrl(StringBuilder urlBuilder, (string key, string value)[] queryParams)
{
    // add script and deploy params to the list
    urlBuilder = this.AppendQueryParam(urlBuilder, ("script", "_restlet.Script"), true);
    urlBuilder = this.AppendQueryParam(urlBuilder, ("deploy", "_restlet.Deploy"));

    foreach(var param in queryParams) {
        urlBuilder = this.AppendQueryParam(urlBuilder, (param.key, param.value));
    }
    return new Uri(urlBuilder.ToString());
}
```

Reply:
StringBuilder is passed by reference, I don't this is a good practice as you are modifying the method parameter.
```
private StringBuilder AppendQueryParam(StringBuilder query, (string key, string value) queryParam, bool isFirstParam = false)
{
    if (isFirstParam)
    {
         query.Append("?");
```

Reply:
@u-pankaj That was the original method signature in `CreateRequestUrl` method. We alternatively create a list of query params in the `AppendQueryParam` method and join them using `&`

Reply:
Good discussion :)

Firstly, @u-pankaj since these are private methods, the fact that we are modifying the param is fine.

Next, @swasti1994 the reason we use `StringBuilder` is to make the generation of this dynamic string as efficient as possible. You have few obvious options:
1. StringBuilder
2. Str1 + str2 + ... (string concatenation)
3. string interpolation.

Think of the implementation of how these things will work:
1. You have a `StringBuilder` which has pre allocated buffer that only needs to be filled up on each call to `Append`.
2. When concatenating you have a 2 strings, both of which the length are known: so you can now allocate a buffer of str1.length + str2.length and copy the chars. Slower than 1, because you had to allocate a new buffer and copy both strings in.
3. How would you (or compiler) go about constructing this string: `$"{queryParam.key}={queryParam.value}"`. String interpolation is basically syntatic sugar for `String.Format`. Firstly, the compiler doesn't know in advance, how complex of a string that it you've given it. So, it cannot pre-allocate buffers. It has to move from substring to substring, variable to variable (and the mix of the two) creating and discarding buffers to generate the string. So, generally:
`StringBuilder` > `Concat` > `Interpolation`.
AFAIK, C# compiler rewrites `Concat` statements using `StringBuilder` that makes even concat perform well. However, given all the different ways in which interpolation can be done complier cannot effectively optimize it.

So, `query.Append($"{queryParam.key}={queryParam.value}");` defeats the purpose of the using the `StringBuilder` in the first place.

Reply:
Now, on the subject of whether we should have an `AppendQueryParam` method:
In many places, extracting a method out makes sense if there's repeated code or when its more readable to do so.

In this case, the proposed is more code than the original:
23 lines vs 17 in the original (would be more if you use the StringBuilder without interpolation)
848 characters vs 520 in the original

This changes, the moment another piece of code needed to do this. Until that happens, it's more efficient to just have the plain original implementation. So, at the moment, I'd say it's better to leave it as it is and refactor if it gets to that.

Reply:
Okay, got it. So, if we need to do this in future, we can append the `&` character, key and value in separate lines instead of interpolating it 👍